pkgs/fc/agent/fc/util/directory.py

```diff
@@ -18,6 +20,40 @@ def __init__(self, url, *args, **kw):
         super(ScreenedProtocolError, self).__init__(url, *args, **kw)
 
 
+RETRY_EXCEPTIONS = (
+    ScreenedProtocolError,
+    xmlrpc.client.Fault,
+    # OSError also covers exceptions from the socket module, ssl.SSLError
+    # and ConnectionError.
+    OSError,
+)
+
+
+class DirectoryAPI(xmlrpc.client.ServerProxy):
+    def __getattr__(self, name):
+        """Magic method dispatcher from ServerProxy with added retry logic."""
+        method = super().__getattr__(name)
+
+        # This function wrapper is needed to make the Stamina decorator work.
+        def wrapper(*args):
+            return method(*args)
+
+        # Stamina uses __qualname__ for retry logging, so let's set it to a
+        # recognizable value.
+        wrapper.__qualname__ = "DirectoryAPI." + name
+
+        retry = stamina.retry(
+            on=RETRY_EXCEPTIONS, wait_exp_base=10, attempts=2
+        )
+        return retry(wrapper)
+
+    def __repr__(self):
+        """ServerProxy.__repr__ leaks the directory password, override it."""
+        host = self._ServerProxy__host.split("@")[1]
+        handler = self._ServerProxy__handler
+        return f"Directory API ({host}{handler})"
+
+
```

Comment:
retry behaviour should be customizable here. default, based on our code base, IMHO should be `off`